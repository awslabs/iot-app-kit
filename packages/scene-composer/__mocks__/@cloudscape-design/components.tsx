import React from 'react';
import mockComponent from '../mockComponent';

const cloudscapeRaw: any = jest.createMockFromModule('@cloudscape-design/components');

const cloudscape = Object.keys(cloudscapeRaw).reduce((acc, comp) => {
  if (!comp.startsWith('_')) {
    acc[comp] = mockComponent(comp);
  } else {
    acc[comp] = cloudscapeRaw[comp];
  }
  return acc;
}, {});

module.exports = {
  ...cloudscape,
  ButtonDropdown: ({ items = [], onItemClick, ...props }: any) => {
    return (
      <ul data-mocked='ButtonDropdown' {...props}>
        { items.map(({ id, text, iconSvg, ...itemProps }) => {
          return (<li id={id} {...itemProps} onClick={() => onItemClick({ detail: { id }})}>{iconSvg}{text}</li>)
        })}
      </ul>
    )
  }
};
