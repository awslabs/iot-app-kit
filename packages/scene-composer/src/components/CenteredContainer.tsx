import { Container } from '@awsui/components-react';
import styled from 'styled-components';

export default styled(Container)`
  width: 600px;
  margin: auto;
  flex: none;
  z-index: 1000;
`;
