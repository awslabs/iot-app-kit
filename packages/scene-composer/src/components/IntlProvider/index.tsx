import React, { useEffect, useState } from 'react';
import { defaultsDeep } from 'lodash';
import PropTypes from 'prop-types';
import { IntlProvider } from 'react-intl';

import useLogger from '../../logger/react-logger/hooks/useLogger';
import { defaultMessages, messagesMap } from '../../../translations';

const DEFAULT_LOCALE = 'en-US';

const Provider = ({ locale, children }) => {
  const log = useLogger('IntlProvider');
  const stringLogger = log?.extend('[MISSINGTRANSLATION]');
  const [messages, setMessages] = useState(defaultMessages);
  const [loc, setLoc] = useState('');

  useEffect(() => {
    log?.verbose('i18nFeature is enabled! Detected locale: ', locale);
    setLoc(locale || DEFAULT_LOCALE);
  }, []);

  useEffect(() => {
    if (loc !== '') {
      let lang = loc;
      if (!(lang in messagesMap)) lang = lang.split('-')[0];
      const strings = messagesMap[lang];
      defaultsDeep(strings, defaultMessages);
      setMessages(strings);
    } else {
      setMessages(defaultMessages);
    }
  }, [loc]);

  return (
    <IntlProvider
      locale={loc}
      messages={messages as unknown as Record<string, string>}
      onError={(err) => stringLogger?.error(err as any as string)}
    >
      {children}
    </IntlProvider>
  );
};
Provider.propTypes = {
  children: PropTypes.node.isRequired,
  locale: PropTypes.string,
};
export default Provider;
