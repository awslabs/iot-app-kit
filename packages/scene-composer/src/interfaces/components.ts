import * as SceneModels from '../models/SceneModels';

/************************************************
 * Components
 ************************************************/

import { DistanceUnit, INavLink, IValueDataBinding, Vector3 } from './dataTypes';

export enum KnownComponentType {
  ModelRef = 'ModelRef',
  SubModelRef = 'SubModelRef',
  Camera = 'Camera',
  Light = 'Light',

  Tag = 'Tag',
  ModelShader = 'ModelShader',
  MotionIndicator = 'MotionIndicator',
  DataOverlay = 'DataOverlay',
}

export interface ISceneComponent {
  // ref is optional when used in APIs to create components
  ref?: string;
  type: KnownComponentType | string;
}

export interface IModelRefComponent extends ISceneComponent {
  uri: string;
  modelType: string;

  unitOfMeasure?: DistanceUnit;
  localScale?: Vector3;
  castShadow?: boolean;
  receiveShadow?: boolean;
}

export interface ISubModelRefComponent extends ISceneComponent {
  parentRef?: string;
  selector: string | number;
}

export interface IAnchorComponent extends ISceneComponent {
  icon?: string;
  valueDataBinding?: IValueDataBinding;
  ruleBasedMapId?: string;
  navLink?: INavLink;
  offset?: Vector3;
}

/**
 * Default Anchor status strings
 */
export enum DefaultAnchorStatus {
  Info = 'Info',
  Warning = 'Warning',
  Error = 'Error',
  Video = 'Video',
}

/**
 * Selected Anchor string
 */
export const SelectedAnchor = 'Selected';

/**
 * Additional Tag Component Data to be given when a tag node is clicked or changed
 */
export interface ITagData {
  navLink?: INavLink;
  dataBindingContext?: unknown;
}

/**
 * Type that can be represented by different additional component data types such as ITagData | IFutureComponentData
 */
export type AdditionalComponentData = ITagData;

/**
 * Callback signature for selection of with Widgets.
 */
export interface ISelectionChangedEvent {
  componentTypes: KnownComponentType[] | string[];
  nodeRef?: string;
  additionalComponentData?: AdditionalComponentData[];
}

/**
 * Callback signature for interaction with Widgets.
 */
export interface IWidgetClickEvent {
  componentTypes: KnownComponentType[] | string[];
  nodeRef: string;
  additionalComponentData?: AdditionalComponentData[];
}

export type WidgetClickEventCallback = (event: IWidgetClickEvent) => void;

export type SelectionChangedEventCallback = (event: ISelectionChangedEvent) => void;

export interface ILightComponent extends ISceneComponent, SceneModels.Component.Light {}

export type CameraType = 'Perspective' | 'Orthographic';

export interface ICameraBasics {
  cameraType: CameraType;
  fov?: number;
  near: number;
  far: number;
  zoom: number;
}

export interface CameraSettings extends ICameraBasics {
  transform: SceneModels.Transform;
}

export interface ICameraComponent extends ISceneComponent, ICameraBasics {}

export interface IColorOverlayComponent extends ISceneComponent {
  valueDataBinding?: IValueDataBinding;
  ruleBasedMapId?: string;
}

export interface IMotionIndicatorComponent extends ISceneComponent, SceneModels.Component.MotionIndicator {}

export interface IDataOverlayComponent extends ISceneComponent, SceneModels.Component.DataOverlay {}
