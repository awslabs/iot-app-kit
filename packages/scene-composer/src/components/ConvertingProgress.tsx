import { Box, Container, ProgressBar } from '@cloudscape-design/components';
import React from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components';
import * as awsui from '@cloudscape-design/design-tokens';

const ConvertingProgressContainer = styled(Container)`
  width: 100%;
`;

const ConvertingProgressDescriptionBox = styled(Box)`
  color: ${awsui.colorTextBodySecondary}!important;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const ConvertingProgress: React.FC<{ total: number; converted: number }> = ({ total, converted }) => {
  const { formatMessage } = useIntl();

  return (
    <ConvertingProgressContainer data-testid='progress-container'>
      {total > 0 && (
        <ProgressBar
          status='in-progress'
          value={(converted / total) * 100}
          additionalInfo={
            <ConvertingProgressDescriptionBox>
              {formatMessage(
                {
                  defaultMessage: '{convertedNumber} out of {totalNumber} converted',
                  description: 'Progress Bar description',
                },
                { convertedNumber: converted, totalNumber: total },
              )}
            </ConvertingProgressDescriptionBox>
          }
          label={formatMessage({ defaultMessage: 'Converting nodes', description: 'Progress Bar label' })}
        />
      )}
    </ConvertingProgressContainer>
  );
};
