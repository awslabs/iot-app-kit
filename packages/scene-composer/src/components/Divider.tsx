import styled from 'styled-components';
import * as awsui from '@cloudscape-design/design-tokens';

export const Divider = styled.hr`
  height: 1px;
  border-width: 0px;
  background-color: ${awsui.colorBorderDividerDefault};
`;
