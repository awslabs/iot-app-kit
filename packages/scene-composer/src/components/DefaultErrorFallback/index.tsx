import React, { FC } from 'react';
import { Header, TextContent } from '@awsui/components-react';

import useLifecycleLogging from '../../logger/react-logger/hooks/useLifecycleLogging';
import { ERROR_MESSAGE_DICT, SceneComposerRuntimeError } from '../../common/errors';
import { StaticLayout } from '../../layouts/StaticLayout';
import CenteredContainer from '../CenteredContainer';

interface ErrorFallbackProps {
  error?: Error;
}

const DefaultErrorFallback: FC<ErrorFallbackProps> = ({ error }) => {
  const log = useLifecycleLogging('DefaultErrorFallback');

  let errorTitle: string;
  let errorMessage: string;

  if (error instanceof SceneComposerRuntimeError) {
    errorTitle = 'Error';
    errorMessage = error.details.message;
    log?.error(errorMessage, error.innerError as Error);
  } else {
    errorTitle = 'Error';
    if (error instanceof Error) {
      errorMessage = error.message;
    } else if (typeof error === 'string') {
      errorMessage = error;
    } else {
      errorMessage = ERROR_MESSAGE_DICT.UNKNOWN_ERROR;
    }
  }

  return (
    <StaticLayout
      showModal={true}
      modalContent={
        <CenteredContainer header={<Header variant='h2'>{errorTitle}</Header>}>
          <TextContent>
            <p>{errorMessage}</p>
          </TextContent>
        </CenteredContainer>
      }
    />
  );
};

export default DefaultErrorFallback;
