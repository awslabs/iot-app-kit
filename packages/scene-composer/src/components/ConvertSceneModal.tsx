import { isEmpty } from 'lodash';
import { Alert, Box, Button, Header, SpaceBetween } from '@awsui/components-react';
import React, { useCallback, useContext, useState } from 'react';
import { useIntl } from 'react-intl';

import { sceneComposerIdContext } from '../common/sceneComposerIdContext';
import { ISceneNodeInternal, useStore } from '../store';
import { KnownSceneProperty } from '../interfaces';
import {
  checkIfEntityAvailable,
  convertAllNodesToEntities,
  createSceneRootEntity,
  prepareWorkspace,
  staticNodeCount,
} from '../utils/entityModelUtils/sceneUtils';
import { createLayer } from '../utils/entityModelUtils/sceneLayerUtils';
import { LayerType } from '../common/entityModelConstants';
import { getGlobalSettings } from '../common/GlobalSettings';

import CenteredContainer from './CenteredContainer';
import { ConvertingProgress } from './ConvertingProgress';

interface ConvertProgress {
  total: number;
  converted: number;
  failedNodes: ISceneNodeInternal[];
  succeededNodes: ISceneNodeInternal[];
}
interface ConvertResult {
  failedNodes?: ISceneNodeInternal[];
  succeededNodes?: ISceneNodeInternal[];
  errorMessage?: string;
}

const ConvertSceneModal: React.FC = () => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const setConvertSceneModalVisibility = useStore(sceneComposerId)((state) => state.setConvertSceneModalVisibility);
  const document = useStore(sceneComposerId)((state) => state.document);

  const setSceneProperty = useStore(sceneComposerId)((state) => state.setSceneProperty);
  const updateSceneNodeInternal = useStore(sceneComposerId)((state) => state.updateSceneNodeInternal);
  const getObject3DBySceneNodeRef = useStore(sceneComposerId)((state) => state.getObject3DBySceneNodeRef);

  const sceneRootId = useStore(sceneComposerId)((state) =>
    state.getSceneProperty<string>(KnownSceneProperty.SceneRootEntityId),
  );
  const sceneLayerIds = useStore(sceneComposerId)((state) =>
    state.getSceneProperty<string[]>(KnownSceneProperty.LayerIds),
  );

  const { formatMessage } = useIntl();

  const [progress, setProgress] = useState<ConvertProgress | undefined>();
  const [result, setResult] = useState<ConvertResult | undefined>();

  const onClose = useCallback(() => {
    setConvertSceneModalVisibility(false);
    setResult(undefined);
    setProgress(undefined);
  }, [setConvertSceneModalVisibility]);

  const onConfirm = useCallback(async () => {
    try {
      const progress: ConvertProgress = {
        total: staticNodeCount(document.nodeMap),
        converted: 0,
        succeededNodes: [],
        failedNodes: [],
      };
      setProgress(progress);

      let layerId = sceneLayerIds?.at(0);
      let rootId = sceneRootId;
      const sceneMetadataModule = getGlobalSettings().twinMakerSceneMetadataModule;
      const layerName = `${sceneMetadataModule?.getSceneId()}_Default`;

      if (sceneMetadataModule) {
        // Before backend can create the new default roots, the client side code will
        // create them temporarily here.
        await prepareWorkspace(sceneMetadataModule);

        // Create default layer and default scene root node
        if (!layerId || isEmpty(layerId) || !(await checkIfEntityAvailable(layerId, sceneMetadataModule))) {
          const layer = await createLayer(layerName, LayerType.Relationship);
          layerId = layer?.entityId;
          setSceneProperty(KnownSceneProperty.LayerIds, [layerId]);
        }
        if (!sceneRootId || isEmpty(sceneRootId) || !(await checkIfEntityAvailable(sceneRootId, sceneMetadataModule))) {
          const root = await createSceneRootEntity();
          rootId = root?.entityId;
          setSceneProperty(KnownSceneProperty.SceneRootEntityId, rootId);
        }
      }

      if (rootId && layerId) {
        convertAllNodesToEntities({
          document,
          sceneRootEntityId: rootId,
          layerId,
          getObject3DBySceneNodeRef,
          onSuccess: (convertedNode) => {
            progress.converted++;
            progress.succeededNodes.push(convertedNode);

            if (progress.converted === progress.total) {
              setProgress(undefined);
              setResult({ failedNodes: progress.failedNodes, succeededNodes: progress.succeededNodes });
            } else {
              setProgress(progress);
            }
            updateSceneNodeInternal(convertedNode.ref, convertedNode, false, true);
          },
          onFailure: (convertedNode) => {
            progress.converted++;
            progress.failedNodes.push(convertedNode);

            if (progress.converted === progress.total) {
              setProgress(undefined);
              setResult({ failedNodes: progress.failedNodes, succeededNodes: progress.succeededNodes });
            } else {
              setProgress(progress);
            }
          },
        });
      }
    } catch (e) {
      setProgress(undefined);
      setResult({ errorMessage: (e as Error).message });
    }
  }, [document, sceneLayerIds, sceneRootId, setSceneProperty, updateSceneNodeInternal]);

  return (
    <CenteredContainer
      header={
        <Header variant='h2'>
          {formatMessage({ description: 'modal header text', defaultMessage: 'Convert scene' })}
        </Header>
      }
      footer={
        <Box float='right' padding={{ bottom: 's' }}>
          {result ? (
            <Button data-testid='ok-button' onClick={onClose}>
              {formatMessage({ description: 'button label', defaultMessage: 'Ok' })}
            </Button>
          ) : (
            <SpaceBetween size='s' direction='horizontal'>
              <Button data-testid='cancel-button' onClick={onClose}>
                {formatMessage({ description: 'button label', defaultMessage: 'Cancel' })}
              </Button>

              <Button data-testid='confirm-button' variant='primary' onClick={onConfirm} disabled={!!progress}>
                {formatMessage({ description: 'button label', defaultMessage: 'Confirm' })}
              </Button>
            </SpaceBetween>
          )}
        </Box>
      }
    >
      {!progress && !result && (
        <Alert
          type='warning'
          header={formatMessage({ description: 'alert header text', defaultMessage: 'Converting scene' })}
        >
          {formatMessage({
            description: 'alert body text',
            defaultMessage:
              'Proceeding with this action will permanently change your scene. You cannot undo this action',
          })}
        </Alert>
      )}
      {progress && <ConvertingProgress total={progress.total} converted={progress.converted} />}
      {result &&
        (isEmpty(result.failedNodes) && !result.errorMessage ? (
          <Alert
            type='success'
            header={formatMessage({
              description: 'alert header text',
              defaultMessage: 'Convert to entities successful',
            })}
          >
            {formatMessage({
              description: 'alert body text',
              defaultMessage: 'All nodes have been converted successfully',
            })}
          </Alert>
        ) : (
          <Alert
            type='error'
            header={formatMessage({ description: 'alert header text', defaultMessage: 'Convert to entities failed' })}
          >
            {result.errorMessage && <Box>{result.errorMessage}</Box>}

            {!isEmpty(result.failedNodes) && (
              <Box>
                {formatMessage({ description: 'alert body text', defaultMessage: 'Some node conversion failed:' })}
              </Box>
            )}
            {result.failedNodes &&
              result.failedNodes.map((node) => (
                <Box key={node.ref} padding={{ left: 's' }}>
                  {node.name}
                </Box>
              ))}
          </Alert>
        ))}
    </CenteredContainer>
  );
};

ConvertSceneModal.displayName = 'ConvertSceneModal';

export default ConvertSceneModal;
