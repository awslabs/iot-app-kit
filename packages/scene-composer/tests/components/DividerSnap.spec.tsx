import React from 'react';
import renderer from 'react-test-renderer';

import { Divider } from '../../src/components/Divider';

describe('Divider', () => {
  it('should render correctly', () => {
    const container = renderer.create(<Divider />);
    expect(container).toMatchSnapshot();
  });
});
