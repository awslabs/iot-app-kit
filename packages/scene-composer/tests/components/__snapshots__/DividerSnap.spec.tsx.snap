// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider should render correctly 1`] = `
.c0 {
  height: 1px;
  border-width: 0px;
  background-color: colorBorderDividerDefault;
}

<hr
  className="c0"
/>
`;
